File: sim/common/debug_gen.rb
#!/usr/bin/env ruby

input_label = ARGV[0].to_i
input_name  = ARGV[1].to_i

def hex_of_floatfile(path)
  float = File.open(path).readlines.map {|line| line.to_f * 256}
  float.map {|num| format("%.4x", num.to_i & 0xffff)}
end

input   = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/mnist/test/#{input_label}/img#{input_name}.dat")
weight0 = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/lenet/conv0/W.dat")
bias0   = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/lenet/conv0/b.dat")
weight1 = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/lenet/conv1/W.dat")
bias1   = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/lenet/conv1/b.dat")

weight2 = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/lenet/full2/W.dat")
bias2   = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/lenet/full2/b.dat")
weight3 = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/lenet/full3/W.dat")
bias3   = hex_of_floatfile("/home/work/takau/2.mlearn/models_chainer/lenet/full3/b.dat")

File.write("../../data/common/#{input_label}_img#{input_name}.dat", input.join("\n"))
File.write("../../data/common/W_conv0.dat", weight0.join("\n"))
File.write("../../data/common/b_conv0.dat", bias0.join("\n"))
File.write("../../data/common/W_conv1.dat", weight1.join("\n"))
File.write("../../data/common/b_conv1.dat", bias1.join("\n"))

File.write("../../data/common/W_full2.dat", weight2.join("\n"))
File.write("../../data/common/b_full2.dat", bias2.join("\n"))
File.write("../../data/common/W_full3.dat", weight3.join("\n"))
File.write("../../data/common/b_full3.dat", bias3.join("\n"))

File: sim/common/debug_gen.py
#!/usr/bin/env python3

import argparse

import os
from os.path import join, exists

import numpy as np

BASE_DIR = "/home/work/takau/2.mlearn/models_chainer"
_BASE_DIR = "../../data/common"
PARAM_DIR = join(BASE_DIR, "lenet")
INPUT_DIR = join(BASE_DIR, "mnist", "test")

def hex_of_float(path):
    floatfile = np.loadtxt(path).astype(np.float) * 256
    fixed = np.around(floatfile).astype(np.int) & 0xffff
    return list(map("{:04x}".format, fixed))

def make(src, dst):
    param = hex_of_float(src)
    with open(dst, "w") as f:
        f.write("\n".join(param)+"\n")

def optparse():
    parser = argparse.ArgumentParser(description="dump parameters for lenet")

    parser.add_argument("input_label")
    parser.add_argument("input_name")

    return parser.parse_args()

def main():
    args = optparse()

    if not exists(_BASE_DIR):
        os.makedirs(_BASE_DIR)

    make(join(INPUT_DIR, args.input_label, f"img{args.input_name}.dat"),
         join(_BASE_DIR, f"{args.input_label}_img{args.input_name}.dat"))

    for layer in os.listdir(PARAM_DIR):
        for Type in ["W", "b"]:
            make(join(PARAM_DIR, layer, f"{Type}.dat"),
                 join(_BASE_DIR, f"{Type}_{layer}.dat"))

if __name__ == "__main__":
    main()

